git-ftp

```diff
@@ -498,7 +497,13 @@ upload_local_sha1() {
 	write_log "Uploading commit log to $REMOTE_BASE_URL_DISPLAY/${REMOTE_PATH}$DEPLOYED_SHA1_FILE."
 	if [ $DRY_RUN -ne 1 ]; then
 		echo "$LOCAL_SHA1" | upload_file - "$DEPLOYED_SHA1_FILE"
-		check_exit_status "Could not upload." "$ERROR_UPLOAD"
+		local EXIT_CODE=$?
+		if [ "$EXIT_CODE" -eq 67 ]; then
```

Comment:
Is this now the third place with the same code? Maybe we should extract it into a function.

Reply:
Yes. This a thought I also had.

I think a function like `check_exit_status` that process more than 2 parameters:
- status code variable (instead of `$?`)
- code to check for (instead of `-ne 0` use `-eq $2`)
- error message (as now)
- exit code to produce (as now)

Is it a idea to use exactly this function but alter the implementation if 4 parameters were given instead of 2 or should we create a new function?

I didn’t implement this cause It should be applied in some places as well I did not want change in this pull request.

Reply:
> I didn’t implement this cause It should be applied in some places as well I did not want change in this pull request.

Ok. Maybe I can introduce this here and we can use it in other case later.

Reply:
Interesting idea. So you would basically do this, right?
```diff
- [ "$EXIT_CODE" -eq 67 ] && print_error_and_die "Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$ERROR_UPLOAD"
+ check_code_for_error "$EXIT_CODE" 67  "Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$ERROR_UPLOAD"
```

It doesn't improve the code much and we still have the repetition. I think we need a method to check curl exit codes:

```sh
check_curl_status() {
  local curl_exit_status=$?
  local unknown_error_message=$1
  local error_code=$2
  case $? in
    0) ;;
    67) print_error_and_die "Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$error_code" ;;
    *) print_error_and_die "Could not upload." "$error_code"
}
```

Maybe that's not the best code yet. You probably get the idea. Maybe we need first a function that maps the curl error code to a good error message and if it returns an error message we then exit:

```sh
message="$(check_curl_status $?)"
if [ -n "$message" ]; then
  print_error_and_die "Could not get last commit: $message exiting..." $ERROR_DOWNLOAD"
fi
```

Reply:
Yes. Except for the function name, this is exactly my idea.

Oh. yes. I got the point. I like your idea, but it would not be that flexible. The function `check_code_for_error` can be used in other cases than curl where we have to check for exit codes. 

But we could do one thing and not lose the other.

And the idea with the message mapping is great as well, but I’d do it the other way round. It could look like this:

```sh
check_curl_status $? "Could not get last commit"

# Checks if a given curl exit code was successful and if not exits with a nice error message (in some cases)
#
# $1 - exit code to check
# $2 - message: a custom message that is put in front of the error message
# $3 - error code: code to exit with if $1 is bad
#
check_curl_status() {
  case $1 in
    0) ;;
    67) print_error_and_die "$2: Can't access remote '$REMOTE_BASE_URL_DISPLAY'. Failed to log in. Correct user and password? exiting..." "$3" ;;
    *) print_error_and_die "$2" "$3"
}
```
This way the code where the check is called becomes cleaner and we don’t need an additional if there. In addition this would make this function similar to the existing behaviour of `check_exit_status` since this exits if the status is bad, too.

And in addition to this we can also provide the other function(s) to use it in other case than curl where we need a more flexible approach:
```sh
# Checks if a given exit code equals a given value
#
# $1 - exit code to check
# $2 - code to test against
# $3 - error message
# $4 - error code to produce
#
check_exit_status_equals() {
	if [ $1 -eq $2 ]; then
		print_error_and_die "$3, exiting..." "$4"
	fi
}

# Checks if a given exit code not equals a given value
#
# $1 - exit code to check
# $2 - code to test against
# $3 - error message
# $4 - error code to produce
#
check_exit_status_not_equals() {
	if [ $1 -ne $2 ]; then
		print_error_and_die "$3, exiting..." "$4"
	fi
}
```

Reply:
Yes, great. Go ahead!